File: apps/proxy/internal/proxy/ratelimit.go
package proxy

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/redis/go-redis/v9"

	"github.com/opendum/opendum/apps/proxy/internal/auth"
)

const apiKeyRateLimitPrefix = "opendum:api-key-rl"

type apiKeyRateLimitResult struct {
	Allowed           bool
	RetryAfterSeconds int
	ExceededWindow    string
	Limit             int
	Current           int
}

const apiKeyRateLimitLua = `
local n = tonumber(ARGV[1])
for i = 1, n do
  local offset = (i - 1) * 3
  local limit = tonumber(ARGV[2 + offset])
  local current = tonumber(redis.call('GET', KEYS[i]) or '0') or 0
  if current >= limit then
    return {0, i, current}
  end
end
for i = 1, n do
  local offset = (i - 1) * 3
  local ttl = tonumber(ARGV[3 + offset])
  local count = redis.call('INCR', KEYS[i])
  if count == 1 then
    redis.call('EXPIRE', KEYS[i], ttl)
  end
end
return {1}
`

var apiKeyRateLimitScript = redis.NewScript(apiKeyRateLimitLua)

func (s *Service) checkAndIncrementAPIKeyRateLimit(ctx context.Context, apiKeyID, model string, rules []auth.RateLimitRule) (apiKeyRateLimitResult, error) {
	rule, ok := s.matchRateLimitRule(model, rules)
	if !ok {
		return apiKeyRateLimitResult{Allowed: true}, nil
	}

	type limitSpec struct {
		window string
		limit  int
		label  string
		secs   int
	}
	limits := []limitSpec{}
	if rule.PerMinute != nil {
		limits = append(limits, limitSpec{window: "min", limit: *rule.PerMinute, label: "minute", secs: 60})
	}
	if rule.PerHour != nil {
		limits = append(limits, limitSpec{window: "hour", limit: *rule.PerHour, label: "hour", secs: 3600})
	}
	if rule.PerDay != nil {
		limits = append(limits, limitSpec{window: "day", limit: *rule.PerDay, label: "day", secs: 86400})
	}
	if len(limits) == 0 {
		return apiKeyRateLimitResult{Allowed: true}, nil
	}

	keys := make([]string, 0, len(limits))
	args := []any{len(limits)}
	for _, limit := range limits {
		keys = append(keys, apiKeyWindowKey(apiKeyID, rule.Target, limit.window, limit.secs))
		args = append(args, limit.limit, limit.secs+1, limit.label)
	}

	res, err := apiKeyRateLimitScript.Run(ctx, s.redis, keys, args...).Slice()
	if err != nil {
		return apiKeyRateLimitResult{Allowed: true}, err
	}
	if len(res) > 0 && toInt(res[0]) == 1 {
		return apiKeyRateLimitResult{Allowed: true}, nil
	}
	idx := toInt(res[1]) - 1
	if idx < 0 || idx >= len(limits) {
		return apiKeyRateLimitResult{Allowed: true}, nil
	}
	exceeded := limits[idx]
	bucket := windowBucket(exceeded.secs)
	retryAfter := maxInt(1, int(bucket+int64(exceeded.secs)-time.Now().Unix()))
	return apiKeyRateLimitResult{Allowed: false, RetryAfterSeconds: retryAfter, ExceededWindow: exceeded.label, Limit: exceeded.limit, Current: toInt(res[2])}, nil
}

func (s *Service) matchRateLimitRule(model string, rules []auth.RateLimitRule) (auth.RateLimitRule, bool) {
	for _, rule := range rules {
		if rule.TargetType == "model" && rule.Target == model {
			return rule, true
		}
	}
	family := s.registry.ModelFamily(model)
	if family != "" {
		for _, rule := range rules {
			if rule.TargetType == "family" && rule.Target == family {
				return rule, true
			}
		}
	}
	return auth.RateLimitRule{}, false
}

func apiKeyWindowKey(apiKeyID, target, window string, windowSeconds int) string {
	return fmt.Sprintf("%s:%s:%s:%s:%d", apiKeyRateLimitPrefix, apiKeyID, target, window, windowBucket(windowSeconds))
}

func windowBucket(windowSeconds int) int64 {
	now := time.Now().Unix()
	return now - (now % int64(windowSeconds))
}

func toInt(value any) int {
	switch v := value.(type) {
	case int:
		return v
	case int64:
		return int(v)
	case uint64:
		return int(v)
	case string:
		parsed, _ := strconv.Atoi(v)
		return parsed
	default:
		return 0
	}
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: apps/proxy/src/core/ratelimit.ts
import type Redis from "ioredis";
import type { RateLimitRule } from "../auth/types.js";

const apiKeyRateLimitPrefix = "opendum:api-key-rl";

export interface APIKeyRateLimitResult {
  allowed: boolean;
  retryAfterSeconds: number;
  exceededWindow: string;
  limit: number;
  current: number;
}

const apiKeyRateLimitLua = `
local n = tonumber(ARGV[1])
for i = 1, n do
  local offset = (i - 1) * 3
  local limit = tonumber(ARGV[2 + offset])
  local current = tonumber(redis.call('GET', KEYS[i]) or '0') or 0
  if current >= limit then
    return {0, i, current}
  end
end
for i = 1, n do
  local offset = (i - 1) * 3
  local ttl = tonumber(ARGV[3 + offset])
  local count = redis.call('INCR', KEYS[i])
  if count == 1 then
    redis.call('EXPIRE', KEYS[i], ttl)
  end
end
return {1}
`;

export async function checkAndIncrementAPIKeyRateLimit(
  redis: Redis | null,
  apiKeyID: string,
  model: string,
  rules: RateLimitRule[],
  modelFamily: string,
): Promise<APIKeyRateLimitResult> {
  const rule = matchRateLimitRule(model, rules, modelFamily);
  if (!rule) return { allowed: true, retryAfterSeconds: 0, exceededWindow: "", limit: 0, current: 0 };

  interface LimitSpec {
    window: string;
    limit: number;
    label: string;
    secs: number;
  }
  const limits: LimitSpec[] = [];
  if (rule.perMinute !== undefined && rule.perMinute !== null) {
    limits.push({ window: "min", limit: rule.perMinute, label: "minute", secs: 60 });
  }
  if (rule.perHour !== undefined && rule.perHour !== null) {
    limits.push({ window: "hour", limit: rule.perHour, label: "hour", secs: 3600 });
  }
  if (rule.perDay !== undefined && rule.perDay !== null) {
    limits.push({ window: "day", limit: rule.perDay, label: "day", secs: 86400 });
  }
  if (limits.length === 0) return { allowed: true, retryAfterSeconds: 0, exceededWindow: "", limit: 0, current: 0 };
  if (!redis) return { allowed: true, retryAfterSeconds: 0, exceededWindow: "", limit: 0, current: 0 };

  const keys: string[] = [];
  const args: Array<string | number> = [limits.length];
  for (const limit of limits) {
    keys.push(apiKeyWindowKey(apiKeyID, rule.target, limit.window, limit.secs));
    args.push(limit.limit, limit.secs + 1, limit.label);
  }

  let result: Array<unknown>;
  try {
    result = (await redis.eval(apiKeyRateLimitLua, keys.length, ...keys, ...args)) as Array<unknown>;
  } catch {
    return { allowed: true, retryAfterSeconds: 0, exceededWindow: "", limit: 0, current: 0 };
  }
  if (result.length > 0 && toInt(result[0]) === 1) {
    return { allowed: true, retryAfterSeconds: 0, exceededWindow: "", limit: 0, current: 0 };
  }
  const idx = toInt(result[1]) - 1;
  if (idx < 0 || idx >= limits.length) {
    return { allowed: true, retryAfterSeconds: 0, exceededWindow: "", limit: 0, current: 0 };
  }
  const exceeded = limits[idx]!;
  const bucket = windowBucket(exceeded.secs);
  const retryAfter = Math.max(1, bucket + exceeded.secs - Math.floor(Date.now() / 1000));
  return { allowed: false, retryAfterSeconds: retryAfter, exceededWindow: exceeded.label, limit: exceeded.limit, current: toInt(result[2]) };
}

function matchRateLimitRule(model: string, rules: RateLimitRule[], family: string): RateLimitRule | null {
  for (const rule of rules) {
    if (rule.targetType === "model" && rule.target === model) return rule;
  }
  if (family !== "") {
    for (const rule of rules) {
      if (rule.targetType === "family" && rule.target === family) return rule;
    }
  }
  return null;
}

function apiKeyWindowKey(apiKeyID: string, target: string, window: string, windowSeconds: number): string {
  return `${apiKeyRateLimitPrefix}:${apiKeyID}:${target}:${window}:${windowBucket(windowSeconds)}`;
}

function windowBucket(windowSeconds: number): number {
  const now = Math.floor(Date.now() / 1000);
  return now - (now % windowSeconds);
}

function toInt(value: unknown): number {
  if (typeof value === "number") return value;
  if (typeof value === "string") {
    const parsed = Number.parseInt(value, 10);
    return Number.isNaN(parsed) ? 0 : parsed;
  }
  return 0;
}
